rclcpp/src/rclcpp/parameter_events_subscriber.cpp

```diff
@@ -0,0 +1,308 @@
+      return true;
+    }
+    return false;
+  }
+  const ParameterEventsCallbackHandle * const base_;
+};
+
+void
+ParameterEventsSubscriber::remove_parameter_callback(
+  const ParameterEventsCallbackHandle * const handle)
+{
+  std::lock_guard<std::recursive_mutex> lock(mutex_);
+  auto & container = parameter_callbacks_[{handle->parameter_name, handle->node_name}];
+  auto it = std::find_if(
+    container.begin(),
+    container.end(),
+    HandleCompare(handle));
+  if (it != container.end()) {
+    container.erase(it);
+    if (container.empty()) {
+      parameter_callbacks_.erase({handle->parameter_name, handle->node_name});
+      if (should_unsubscribe_to_namespace(split_path(handle->node_name).first)) {
+        remove_namespace_event_subscriber(split_path(handle->node_name).first);
+      }
+    }
+  } else {
+    throw std::runtime_error("Callback doesn't exist");
+  }
+}
+
+bool
+ParameterEventsSubscriber::should_unsubscribe_to_namespace(const std::string & node_namespace)
+{
+  auto it1 = std::find(event_namespaces_.begin(), event_namespaces_.end(), node_namespace);
+  if (it1 != event_namespaces_.end()) {
+    return false;
+  }
+
+  auto it2 = parameter_callbacks_.begin();
+  while (it2 != parameter_callbacks_.end()) {
+    if (split_path(it2->first.second).first == node_namespace) {
+      return false;
+    }
+    ++it2;
+  }
+  return true;
+}
+
+void
+ParameterEventsSubscriber::remove_parameter_callback(
+  const std::string & parameter_name,
+  const std::string & node_name)
+{
+  std::lock_guard<std::recursive_mutex> lock(mutex_);
+  auto full_node_name = resolve_path(node_name);
+  parameter_callbacks_.erase({parameter_name, full_node_name});
+  if (should_unsubscribe_to_namespace(split_path(full_node_name).first)) {
+    RCLCPP_INFO(node_logging_->get_logger(), "Removing namespace event subscription");
+    remove_namespace_event_subscriber(split_path(full_node_name).first);
+  }
+}
+
+bool
+ParameterEventsSubscriber::get_parameter_from_event(
+  const rcl_interfaces::msg::ParameterEvent::SharedPtr event,
+  rclcpp::Parameter & parameter,
+  const std::string parameter_name,
+  const std::string node_name)
+{
+  if (event->node == node_name) {
+    rclcpp::ParameterEventsFilter filter(event, {parameter_name},
+      {rclcpp::ParameterEventsFilter::EventType::NEW,
+        rclcpp::ParameterEventsFilter::EventType::CHANGED});
+    if (!filter.get_events().empty()) {
+      const auto & events = filter.get_events();
+      auto param_msg = events.back().second;
+      parameter = rclcpp::Parameter::from_parameter_msg(*param_msg);
+      return true;
+    }
+  }
+  return false;
+}
+
+rclcpp::Parameter
+ParameterEventsSubscriber::get_parameter_from_event(
+  const rcl_interfaces::msg::ParameterEvent::SharedPtr event,
+  const std::string parameter_name,
+  const std::string node_name)
+{
+  rclcpp::Parameter p(parameter_name);
+  get_parameter_from_event(event, p, parameter_name, node_name);
```

Comment:
Maybe, it's better to  throw when you can't get the parameter from the event ...

Reply:
Hmm, that may be better than requiring the user to check themselves